### PR TITLE
PDF goto link feature

### PR DESCRIPTION
**Pull request by Technius**
This PR adds support for navigating to links contained in a PDF background file. Implements #1027.

* Alt-clicking on a link (both a URI or a link to another page) will open a popup.
    * For a URI, the popup will show the URI, which can be clicked to open the default application for that URI (e.g., web browser for http/https, email client for mailto).
    * For an internal link, the application will scroll to that page. If the page is missing, the "missing PDF page" dialog will be displayed.

This PR is a rebased version of #3851 which is a rebased version of #2660.

### TODO
* [ ] The changes required to support relative links are rather complicated, and could be moved into a separate PR.
* [x] Use the PDF text selection tools instead of the element selection tools.
* [x] Run IWYU to fix the includes (help appreciated here -- I can't get it to work on my computer)
* [x] Apply my own code suggestions from #3851 
* [x] ~For touchscreen, we could add long-tap to open the link popup.~ (This is somewhat unnecessary. Touchscreen works fine with the PDF text selection tools when touch drawing is enabled).
* [x] Fix macOS build failure

### CODE REVIEW

**Comment by bhennion** [comment]
> * [ ]  Run IWYU to fix the includes (help appreciated here -- I can't get it to work on my computer)

I'll do it once the code is stabilized.

Also, I think we could make the hand tool (which I see as a navigation tool) follow links when clicking (without pressing alt or anything).

**Comment by Technius** [comment]
After some experimentation, I found it intuitive to display the navigation popup if the cursor is positioned above the link when using the PDF text selection tools. This has the benefit of working if you "click" the link with a PDF text selection tool.

https://user-images.githubusercontent.com/1066652/202922045-9de00af2-3ba0-46ae-a2ea-1b44c84850b0.mp4

(In the recording, the cursor is large and the position is off due to some bug in the recording software on Wayland).

File: src/core/gui/PageView.cpp [hunk 1]
```diff
@@ -890,6 +914,94 @@ void XojPageView::drawLoadingPage(cairo_t* cr) {
...
+        }
+
+        GtkWidget* box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 5);
+        GtkWidget* popover = makePopover(rect, box);
+
+        if (dest->isURI()) {
+            std::string uri = dest->getURI();
+            char* uriLabel = g_markup_escape_text(uri.c_str(), -1);
+
+            std::stringstream labelMarkup;
```
Comment (rolandlo) [inline src/core/gui/PageView.cpp:935]: `#include <sstream>`
is missing, I think. That's why it doesn't compile on MacOS.

**Review by bhennion — COMMENTED** [review]
Mostly nitpicks. I still have to try it out.

Tell me once you've made the changes you deam useful, and I'll try and run IWYU on the PR. **Edit**: done

**Comment by bhennion** [comment]
Seems to work as expected (tested only on linux).

Two suggestions:

- It took me a couple of minutes to understand you had to select the pdf selection tools to make it work. I would add the hand tool to the list.
- The popover for external links could be improved:
    - add a small margin
    - maybe having the link displayed (but not clickable) and then two entries: one "Open link in browser" and one "Copy link".

**Comment by bhennion** [comment]
@Technius I pushed a commit applying IWYU. Pull before modifying :-)

File: src/core/pdf/popplerapi/PopplerGlibPage.cpp [hunk 2]
```diff
@@ -287,3 +293,21 @@ auto PopplerGlibPage::selectTextLines(const XojPdfRectangle& selectRect, XojPdfP
     g_assert_nonnull(region);
     return {.region = xoj::util::CairoRegionSPtr(region, xoj::util::adopt), .rects = textRects};
 }
+
+auto PopplerGlibPage::getLinks() -> std::vector<Link> {
+    std::vector<Link> results;
+    const double height = getHeight();
+
+    GList* links = poppler_page_get_link_mapping(this->page);
+    for (GList* l = links; l != NULL; l = g_list_next(l)) {
```
[the remark this answers is not in the document]
Comment (Technius) [reply src/core/pdf/popplerapi/PopplerGlibPage.cpp:306]: GListView segfaults when I use it in the obvious way:
```
for (auto l* : GListView<PopplerLinkMapping*>(l)) {
```

File: src/core/model/LinkDestination.cpp [hunk 3]
```diff
@@ -83,5 +85,14 @@ void LinkDestination::setChangeTop(double top) {
 }
 
 void LinkDestination::setName(std::string name) { this->name = std::move(name); }
-
-auto LinkDestination::getName() -> std::string { return this->name; }
+auto LinkDestination::getName() -> std::string { return name; }
+
+void LinkDestination::setURI(std::string uri) { this->contents = LinkType{std::move(uri)}; }
+auto LinkDestination::getURI(bool copy) const -> std::optional<std::string> {
```
[the remark this answers is not in the document]
Comment (Technius) [reply src/core/model/LinkDestination.cpp]: I forgot that `std::holds_alternative` exists :sweat_smile: ... I've always been using `std::get_if`. It should be cleaner now.

**Comment by Technius** [comment]
I applied the code cleanup suggestions, will look into the other requests tomorrow.

> It took me a couple of minutes to understand you had to select the pdf selection tools to make it work. I would add the hand tool to the list.

Sorry for the confusion, I forgot to update the PR description.

I'm not entirely convinced that adding the hand tool (with directly tap/click) is desirable:
1. Some people may want the hand tool to only be used for panning, and it could be annoying if someone touches an object that covers up a link but the link is triggered.
2. Automatically opening URLs could be a security problem.
3. If a document contains a lot of links, the user would have to "dodge" the links when panning with the hand tool. This could be very annoying.

Points 1 & 3 can be avoided if we bind the link interaction to a different key/button, but this would have the downside of needing another button. 

> The popover for external links could be improved

These suggestions sound good to me. Actually, I was thinking that we might want to place the link into a scrollable text field/box and impose a maximum length on the link, otherwise a long link could potentially go out of the screen. Putting a copy button next to it would then feel fairly natural (note that with the current setup, you can copy the link by right clicking it and using the displayed menu item).

File: src/core/pdf/popplerapi/PopplerGlibPage.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhennion) [reply src/core/pdf/popplerapi/PopplerGlibPage.cpp:306]: I think it's used `GListView<PopplerLinkMapping>` (no pointer)

**Comment by Technius** [comment]
> If so, then there should be no interaction problems between the panning and links: panning only happens for a drap'n'drop, while link are only followed for clicks/taps. I'd expect something like Okular's hand tool, or every touch-aware webbrowser (tap opens a link, drag'n'drop scrolls).

Oh, I understand better now; this makes sense to me. I'm not too familiar with the input handling code, so it'd be helpful if you could point me to how we can distinguish between click/tap and drag&drop.

> For the possible security risk, either open a popover when an external link is clicked upon, or check what other pdf viewers do. Okular seems to simply follow the links.

I personally prefer the popover option, but either way should not present any challenges.

**Comment by bhennion** [comment]
> Oh, I understand better now; this makes sense to me. I'm not too familiar with the input handling code, so it'd be helpful if you could point me to how we can distinguish between click/tap and drag&drop.

I just pushed a PR with code I wrote a while back. See #4501. This should help with understanding how to detect taps and so on. Note that #4501 is based on #4159, so look only at the last commit of #4501. This last commit might be cherry-pickable to master, don't know, didn't try yet.

> I personally prefer the popover option

Fine by me.

**Comment by Technius** [comment]
I squashed the commits. I checked the diff between the unsquashed version, and the only change is the removal of a redundant include in `LinkDestination.cpp`.

The hand tool interaction will be left to a follow up PR.

Merging in 24 hours if there are no objections.

**Pull request merged**